src/Workspaces/SharedUtilitiesAndExtensions/Compiler/Core/Collections/SegmentedArray`1.cs

```diff
@@ -0,0 +1,358 @@
+        }
+
+        int IList.IndexOf(object value)
+        {
+            for (var i = 0; i < _items.Length; i++)
+            {
+                IList list = _items[i];
+                var index = list.IndexOf(value);
+                if (index >= 0)
+                {
+                    return index + i * s_segmentSize;
+                }
+            }
+
+            return -1;
+        }
+
+        int IList<T>.IndexOf(T value)
+        {
+            for (var i = 0; i < _items.Length; i++)
+            {
+                IList<T> list = _items[i];
+                var index = list.IndexOf(value);
+                if (index >= 0)
+                {
+                    return index + i * s_segmentSize;
+                }
+            }
+
+            return -1;
+        }
+
+        void IList.Insert(int index, object value)
+        {
+            throw new NotSupportedException(CompilerExtensionsResources.NotSupported_FixedSizeCollection);
+        }
+
+        void IList<T>.Insert(int index, T value)
+        {
+            throw new NotSupportedException(CompilerExtensionsResources.NotSupported_FixedSizeCollection);
+        }
+
+        void IList.Remove(object value)
+        {
+            throw new NotSupportedException(CompilerExtensionsResources.NotSupported_FixedSizeCollection);
+        }
+
+        bool ICollection<T>.Remove(T value)
+        {
+            throw new NotSupportedException(CompilerExtensionsResources.NotSupported_FixedSizeCollection);
+        }
+
+        void IList.RemoveAt(int index)
+        {
+            throw new NotSupportedException(CompilerExtensionsResources.NotSupported_FixedSizeCollection);
+        }
+
+        void IList<T>.RemoveAt(int index)
+        {
+            throw new NotSupportedException(CompilerExtensionsResources.NotSupported_FixedSizeCollection);
+        }
+
+        IEnumerator<T> IEnumerable<T>.GetEnumerator()
+            => GetEnumerator();
+
+        IEnumerator IEnumerable.GetEnumerator()
+            => GetEnumerator();
+
+        int IStructuralComparable.CompareTo(object other, IComparer comparer)
+        {
+            if (other is null)
+                return 1;
+
+            if (!(other is SegmentedArray<T> o)
+                || Length != o.Length)
+            {
+                throw new ArgumentException(CompilerExtensionsResources.ArgumentException_OtherNotArrayOfCorrectLength, nameof(other));
+            }
+
+            for (var i = 0; i < Length; i++)
+            {
+                var result = comparer.Compare(this[i], o[i]);
+                if (result != 0)
+                    return result;
+            }
+
+            return 0;
+        }
+
+        bool IStructuralEquatable.Equals(object other, IEqualityComparer comparer)
+        {
+            if (other is null)
+                return false;
+
+            if (!(other is SegmentedArray<T> o))
+                return false;
+
+            if ((object)_items == o._items)
```

Comment:
object cast here seems unnecessary.

Reply:
I wanted to keep the intentional reference equality clear.

Reply:
The diagnostic we have for unnecessary cast already recognizes this case and does not suggest removing it, so this intent is not at risk of being lost.